src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -57,19 +107,78 @@ internal ConfigurationClientManager(IList<ConfigurationClientWrapper> clients)
         {
-            return _clients.Where(client => client.BackoffEndTime <= time).Select(c => c.Client).ToList();
-        }
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+
+            // Principle of refreshing fallback clients:
+            // 1.Perform initial refresh attempt to query available clients on app start-up.
+            // 2.Refreshing when cached fallback clients have expired, FallbackClientRefreshExpireInterval is due.
+            // 3.At least wait MinimalClientRefreshInterval between two attempt to ensure not perform refreshing too often.
+
+            if (_replicaDiscoveryEnabled &&
+                now >= _lastFallbackClientRefresh + MinimalClientRefreshInterval && 
+                (_dynamicClients == null ||
+                now >= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval))
+            {
+                await RefreshFallbackClients(cancellationToken).ConfigureAwait(false);
+
+                _lastFallbackClientRefresh = now;
+            }
 
+            foreach (ConfigurationClientWrapper clientWrapper in _clients)
+            {
+                if (clientWrapper.BackoffEndTime <= now)
+                {
+                    yield return clientWrapper.Client;
+                }
+            }
+
+            // No need to continue if replica discovery is not enabled.
+            if (!_replicaDiscoveryEnabled)
+            {
+                yield break;
+            }
+
+            foreach (ConfigurationClientWrapper clientWrapper in _dynamicClients)
+            {
+                if (clientWrapper.BackoffEndTime <= now)
+                {
+                    yield return clientWrapper.Client;
+                }
+            }
+
+            // All static and dynamic clients exhausted, refresh fallback clients if
+            // minimal client refresh interval is due.
+            if (now >= _lastFallbackClientRefresh + MinimalClientRefreshInterval)
+            {
+                await RefreshFallbackClients(cancellationToken).ConfigureAwait(false);
+
+                _lastFallbackClientRefresh = now;
+            }
+
+            foreach (ConfigurationClientWrapper clientWrapper in _dynamicClients)
+            {
+                if (clientWrapper.BackoffEndTime <= now)
+                {
+                    yield return clientWrapper.Client;
+                }
+            }
+        }
+
         public void UpdateClientStatus(ConfigurationClient client, bool successful)
         {
             if (client == null)
             {
                 throw new ArgumentNullException(nameof(client));
             }
 
-            ConfigurationClientWrapper clientWrapper = _clients.First(c => c.Client.Equals(client));
+            ConfigurationClientWrapper clientWrapper = _clients.FirstOrDefault(c => c.Client.Equals(client));
+
+            if (clientWrapper == null)
+            {
+                clientWrapper = _dynamicClients.First(c => c.Client.Equals(client));

```

Comment:
`_dynamicClients` changes over time. Using `First` rather than `FirstOrDefault` seems dangerous in that case.

Reply:
Good catach, although I can't figure out what exact scenario would be buggy case for using `First`, it's potentially dangerous, I will update to use `FirstOrDefault` and do null check